### PR TITLE
Adapt AList basically

### PR DESCRIPTION
**Pull request by Roozenlz**
完成项目基本功能，后续进行代码重构和测试

```release-note
None
```

### CODE REVIEW

**Review by JohnNiang — COMMENTED** [review]
我这边上传一个普通文本文件似乎会出现以下错误：

```java
2024-07-16T00:13:54.216+08:00 ERROR 94767 --- [ctor-http-nio-1] a.w.r.e.AbstractErrorWebExceptionHandler : [780aca64-2171]  500 Server Error for HTTP POST "/apis/api.console.halo.run/v1alpha1/attachments/upload"

java.lang.RuntimeException: strconv.ParseInt: parsing "": invalid syntax
	at run.halo.alist.AListAttachmentHandler.lambda$upload$1(AListAttachmentHandler.java:78) ~[na:na]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ run.halo.app.security.InitializeRedirectionWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.security.device.DeviceSessionFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthorizationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ExceptionTranslationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ LogoutWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ServerRequestCacheWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityContextServerWebExchangeWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ LogoutPageGeneratingWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AnonymousAuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ TotpAuthenticationFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ReactorContextWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ CsrfWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ CorsWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ HttpHeaderWriterWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ ServerWebExchangeReactorContextWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ org.springframework.security.web.server.WebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.webfilter.AdditionalWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ HTTP POST "/apis/api.console.halo.run/v1alpha1/attachments/upload" [ExceptionHandlingWebHandler]
Original Stack Trace:
		at run.halo.alist.AListAttachmentHandler.lambda$upload$1(AListAttachmentHandler.java:78) ~[na:na]
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:132) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondComplete(MonoFlatMap.java:245) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:305) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:158) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:299) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxFilterFuseable$FilterFuseableConditionalSubscriber.onNext(FluxFilterFuseable.java:337) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.Operators$BaseFluxToMonoOperator.completePossiblyEmpty(Operators.java:2097) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.MonoCollect$CollectSubscriber.onComplete(MonoCollect.java:145) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:415) ~[reactor-netty-core-1.1.20.jar:1.1.20]
		at reactor.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:446) ~[reactor-netty-core-1.1.20.jar:1.1.20]
		at reactor.netty.channel.ChannelOperations.terminate(ChannelOperations.java:500) ~[reactor-netty-core-1.1.20.jar:1.1.20]
		at reactor.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:793) ~[reactor-netty-http-1.1.20.jar:1.1.20]
		at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:114) ~[reactor-netty-core-1.1.20.jar:1.1.20]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[netty-codec-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
		at java.base/java.lang.Thread.run(Thread.java:840) ~[na:na]

```

**Comment by Roozenlz** [comment]
> 我这边上传一个普通文本文件似乎会出现以下错误：
> 
> ```java
> 2024-07-16T00:13:54.216+08:00 ERROR 94767 --- [ctor-http-nio-1] a.w.r.e.AbstractErrorWebExceptionHandler : [780aca64-2171]  500 Server Error for HTTP POST "/apis/api.console.halo.run/v1alpha1/attachments/upload"
> 
> java.lang.RuntimeException: strconv.ParseInt: parsing "": invalid syntax
> 	at run.halo.alist.AListAttachmentHandler.lambda$upload$1(AListAttachmentHandler.java:78) ~[na:na]
> 	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
> Error has been observed at the following site(s):
> 	*__checkpoint ⇢ run.halo.app.security.InitializeRedirectionWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ run.halo.app.security.device.DeviceSessionFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
> 	*__checkpoint ⇢ AuthorizationWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ ExceptionTranslationWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ LogoutWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ ServerRequestCacheWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ SecurityContextServerWebExchangeWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ LogoutPageGeneratingWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ AnonymousAuthenticationWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
> 	*__checkpoint ⇢ TotpAuthenticationFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
> 	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
> 	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ ReactorContextWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ CsrfWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ CorsWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ HttpHeaderWriterWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
> 	*__checkpoint ⇢ ServerWebExchangeReactorContextWebFilter [DefaultWebFilterChain]
> 	*__checkpoint ⇢ org.springframework.security.web.server.WebFilterChainProxy [DefaultWebFilterChain]
> 	*__checkpoint ⇢ run.halo.app.webfilter.AdditionalWebFilterChainProxy [DefaultWebFilterChain]
> 	*__checkpoint ⇢ HTTP POST "/apis/api.console.halo.run/v1alpha1/attachments/upload" [ExceptionHandlingWebHandler]
> Original Stack Trace:
> 		at run.halo.alist.AListAttachmentHandler.lambda$upload$1(AListAttachmentHandler.java:78) ~[na:na]
> 		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:132) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondComplete(MonoFlatMap.java:245) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:305) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:158) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:299) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.FluxFilterFuseable$FilterFuseableConditionalSubscriber.onNext(FluxFilterFuseable.java:337) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.Operators$BaseFluxToMonoOperator.completePossiblyEmpty(Operators.java:2097) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.MonoCollect$CollectSubscriber.onComplete(MonoCollect.java:145) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144) ~[reactor-core-3.6.7.jar:3.6.7]
> 		at reactor.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:415) ~[reactor-netty-core-1.1.20.jar:1.1.20]
> 		at reactor.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:446) ~[reactor-netty-core-1.1.20.jar:1.1.20]
> 		at reactor.netty.channel.ChannelOperations.terminate(ChannelOperations.java:500) ~[reactor-netty-core-1.1.20.jar:1.1.20]
> 		at reactor.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:793) ~[reactor-netty-http-1.1.20.jar:1.1.20]
> 		at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:114) ~[reactor-netty-core-1.1.20.jar:1.1.20]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[netty-codec-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
> 		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.111.Final.jar:4.1.111.Final]
> 		at java.base/java.lang.Thread.run(Thread.java:840) ~[na:na]
> ```

你试一下直接在AList 网页端上传呢，我这边看到是AList接口返回的报错信息为 strconv.ParseInt: parsing "": invalid syntax ，请求响应码不为200

**Comment by Roozenlz** [comment]
可能部分云盘不支持上传该类型的文件

**Comment by JohnNiang** [comment]
复现步骤：

0. 安装 Alist 最新版
1. 在 Alist 侧创建 `Local` 存储，并挂载到 `/halo`。
2. 在 Halo 侧创建新的存储策略，并验证通过。
3. 上传任意文件即可看到以上错误。

File: src/main/java/run/halo/alist/AListAttachmentHandler.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    @Override
+    public Mono<Attachment> upload(UploadContext uploadContext) {
+        return Mono.just(uploadContext).filter(context -> this.shouldHandle(context.policy()))
+            .map(UploadContext::configMap)
+            .map(this::getProperties)
+            .flatMap(this::auth)
+            .flatMap(token -> webClient.put()
+                .uri("/api/fs/put")
+                .header("Authorization", token)
+                .header("File-Path", UriUtils.encodePath(
```
Comment (JohnNiang) [inline src/main/java/run/halo/alist/AListAttachmentHandler.java:64]: 从源码发现，这里可能需要传入 `Content-Length` 请求头。请参考：<https://github.com/alist-org/alist/blob/a93937f80da8011c2dbd6464967d190b20ed8301/server/handles/fsup.go#L45-L47>。

Comment (Roozenlz) [reply src/main/java/run/halo/alist/AListAttachmentHandler.java:64]: 不过UploadContext没找到获取文件大小的方法，我了解到的方法是需要阻塞调用才能得到文件大小

Comment (JohnNiang) [reply src/main/java/run/halo/alist/AListAttachmentHandler.java]: 从 https://datatracker.ietf.org/doc/html/rfc7578#page-7 了解到，file.headers() 中应该是无法获取到 Content-Length 信息的。所以我们只能通过重复读取 file.content() 的方式来获取文件大小。如果是上传大文件的话，可能会比较慢，不过也是可以接受的。

File: src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+ * @version 1.0
+ * 2024/7/3
+ */
+@Slf4j
+@Extension
+@Component
+public class AListAttachmentHandler implements AttachmentHandler {
+
+    @Autowired
+    private ReactiveExtensionClient client;
```
Comment (JohnNiang) [inline src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java:62]: 推荐使用构造函数的方式引入依赖，并加上 final 关键字。

File: src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+    private AListProperties properties = null;
```
Comment (JohnNiang) [inline src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java:64]: 可以考虑在 afterProperties 中初始化 properties。同时可以监听事件 `PluginConfigUpdatedEvent` 来更新 properties。

File: src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+            .map(UploadContext::configMap)
+            .map(this::getProperties)
+            .flatMap(this::auth)
+            .flatMap(token -> file.content()
+                .map(dataBuffer -> (long) dataBuffer.readableByteCount())
+                .reduce(Long::sum)
+                .flatMap(fileSize -> webClients.get(properties.getSite())
+                    .put()
+                    .uri("/api/fs/put")
+                    .header("Authorization", token)
```
Comment (JohnNiang) [inline src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java:87]: 请求头字符串建议用 `org.springframework.http.HttpHeaders#AUTHORIZATION` 替代。

File: src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java (hunk 3 above)
Comment (Roozenlz) [reply src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java:64]: 我这里的properties不是插件配置，而且存储策略的配置信息，每个存储配置都需要在UploadContext获取，对于多个存储策略，初始化不了吧，不过可以加个缓存，另外PluginConfigUpdatedEvent这个事件用不了吧

**Review by JohnNiang — APPROVED** [review]
/retitle Adapt AList basically
/lgtm

Thanks for your contribution!

File: src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+                    }))
+            )
+            .map(fileSize -> {
+                var metadata = new Metadata();
+                metadata.setName(UUID.randomUUID().toString());
+                metadata.setAnnotations(
+                    Map.of(Constant.EXTERNAL_LINK_ANNO_KEY,
+                        UriUtils.encodePath(
+                            properties.getSite() + "/d" + properties.getPath() + "/"
+                                + file.name(),
```
Comment (JohnNiang) [inline src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java:123]: 这里建议使用 `file.filename()`。

`file.name()` 返回的结果是表单参数名，而不是真实的文件名。

File: src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                                file.name());
+                            return Mono.just(fileSize);
+                        }
+                        return Mono.error(new AListRequestErrorException(response.getMessage()));
+                    }))
+            )
+            .map(fileSize -> {
+                var metadata = new Metadata();
+                metadata.setName(UUID.randomUUID().toString());
+                metadata.setAnnotations(
```
Comment (JohnNiang) [inline src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java:119]: 建议在这里把 File-Path 写入到 annotations（`alist.storage.halo.run/file-path`）中，这样方便接下来调用 getPermalink 和 delete，不建议直接使用 displayName，因为该名称可随时被更改。

File: src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java (hunk 5 above)
Comment (Roozenlz) [reply src/main/java/run/halo/alist/endpoint/AListAttachmentHandler.java:123]: 这里已经修改成空字符串了，我发现不设置Constant.EXTERNAL_LINK_ANNO_KEY，就不执行getPermalink，所以还是给了个空值

**Pull request merged**